src/batch.rs

```diff
@@ -0,0 +1,187 @@
+use crate::Agate;
+use crate::{entry::Entry, ops::transaction::Transaction};
+use crate::{Error, Result};
+
+use bytes::Bytes;
+use std::sync::Arc;
+
+/// WriteBatch helps write multiple entries to database
+pub struct WriteBatch {
+    txn: Transaction,
+    core: Arc<crate::db::Core>,
+    err: Option<Error>,
+
+    is_managed: bool,
+    commit_ts: u64,
+    // TODO: Add finished and support concurrency.
+}
+
+impl Agate {
+    /// Creates a new [`WriteBatch`]. This provides a way to conveniently do a lot of writes,
+    /// batching them up as tightly as possible in a single transaction, thus achieving good
+    /// performance. This API hides away the logic of creating and committing transactions.
+    /// Due to the nature of SSI guaratees provided by Agate, blind writes can never encounter
+    /// transaction conflicts.
+    pub fn new_write_batch(&self) -> WriteBatch {
+        if self.core.opts.managed_txns {
+            panic!("Cannot use new_write_batch in managed mode. Use new_write_batch_at instead");
+        }
+
+        WriteBatch {
+            txn: self.core.new_transaction(true, true),
+            core: self.core.clone(),
+            err: None,
+            is_managed: false,
+            commit_ts: 0,
+        }
+    }
+
+    /// Similar to `new_write_batch` but it allows user to set the commit timestamp.
+    pub fn new_write_batch_at(&self, commit_ts: u64) -> WriteBatch {
+        if !self.core.opts.managed_txns {
+            panic!(
+                "Cannot use new_write_batch_at in non-managed mode. Use new_write_batch instead."
+            );
+        }
+
+        let mut txn = self.core.new_transaction(true, true);
+        txn.commit_ts = commit_ts;
+        WriteBatch {
+            txn,
+            core: self.core.clone(),
+            err: None,
+            is_managed: true,
+            commit_ts,
+        }
+    }
+}
+
+impl WriteBatch {
+    pub fn set(&mut self, key: Bytes, value: Bytes) -> Result<()> {
+        let entry = Entry::new(key, value);
+        self.set_entry(entry)
+    }
+
+    pub fn set_entry(&mut self, entry: Entry) -> Result<()> {
+        self.handle_entry(entry)
+    }
+
+    fn handle_entry(&mut self, entry: Entry) -> Result<()> {
+        if let Err(err) = self.txn.set_entry(entry.clone()) {
+            if !matches!(err, crate::Error::TxnTooBig) {
+                self.err = Some(err.clone());
+                return Err(err);
+            }
+        } else {
+            return Ok(());
+        }
+
+        self.commit()?;
+
+        if let Err(err) = self.txn.set_entry(entry) {
```

Comment:
I don't understand, why `set_entry` twice

Reply:
At first `set_entry`, we may get `TxnTooBig` error. In that case, we should commit and then `set_entry` again.